akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala

```diff
@@ -170,16 +172,24 @@ import scala.compat.java8.FutureConverters._
     }
 
     (stageLogic, new SourceQueueWithComplete[T] {
+      @volatile
```

Comment:
do we need volatile here?
`complete()` and `fail()` can basically be called from multiple threads.